src/main/java/hq/RestoreUtils.java

```
@@ -25,11 +34,60 @@ public static UserSqlSandbox restoreUser(String username, String restorePayload)
-        UserSqlSandbox mSandbox = SqlSandboxUtils.getStaticStorage(restoreRequest.getUsername());
-        PrototypeFactory.setStaticHasher(new ClassNameHasher());
-        ParseUtilsHelper.parseXMLIntoSandbox(restorePayload, mSandbox);
-        return mSandbox;
+    /**
+     * Build a form definition and load a particular form instance into it.
+     * The FormDef object returned isn't initialized, and hence will not have
+     * 'instance(...)' data set.
+     *
+     * @param formInput     XML stream of the form definition
+     * @param instanceInput XML stream of an instance of the form
+     * @return The form definition with the given instance loaded. Returns null
+     * if the instance doesn't match the form provided.
+     * @throws IOException thrown when XML input streams aren't successfully
+     *                     parsed
+     */
+    public static FormDef loadInstance(InputStream formInput,
+                                       InputStream instanceInput)
+            throws IOException {
+        FormDef formDef;
+        FormInstance savedModel;
+        FormEntryModel entryModel;
+
+        try {
+            formDef = XFormUtils.getFormFromInputStream(formInput);
+        } catch (XFormParseException e) {
+            throw new IOException(e.getMessage());
+        }
+
+        savedModel = XFormParser.restoreDataModel(instanceInput, null);
+
+        // get the root of the saved and template instances
+        TreeElement savedRoot = savedModel.getRoot();
+        TreeElement templateRoot =
+                formDef.getInstance().getRoot().deepCopy(true);
+
+        entryModel = new FormEntryModel(formDef);
+
+        // weak check for matching forms
+        if (!savedRoot.getName().equals(templateRoot.getName()) ||
+                savedRoot.getMult() != 0) {
+            System.out.println("Instance and template form definition don't match");
+            return null;
```

Should maybe raise an exception instead since this is an error condition